### DISCUSSION TITLE
Is the RCT_EXTERN_METHOD not working in new arch?

### DISCUSSION THREAD

**Discussion by XuNeal** [mateusz1913/rnbridgingtutorial#36] [Q&A] [answered]
I impl my biz in Swift, and export the api in *Exports.m like below

``` Objective-C
@interface RCT_EXTERN_MODULE(ClassName, NSObject)

RCT_EXTERN_METHOD(doSomething:(NSDictionary *)map resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end 
```

I don't need to write a wrap method in OC handly, which RCT_EXTERN_METHOD helps me do that. The comment of RCT_EXTERN_MODULE tell all things about that.
```
/**
 * Use this macro in a private Objective-C implementation file to automatically
 * register an external module with the bridge when it loads. This allows you to
 * register Swift or private Objective-C classes with the bridge.
 *
 * For example if one wanted to export a Swift class to the bridge:
 *
 * MyModule.swift:
 *
 *   @objc(MyModule) class MyModule: NSObject {
 *
 *     @objc func doSomething(string: String! withFoo a: Int, bar b: Int) { ... }
 *
 *   }
 *
 * MyModuleExport.m:
 *
 *   #import <React/RCTBridgeModule.h>
 *
 *   @interface RCT_EXTERN_MODULE(MyModule, NSObject)
 *
 *   RCT_EXTERN_METHOD(doSomething:(NSString *)string withFoo:(NSInteger)a bar:(NSInteger)b)
 *
 *   @end
 *
 * This will now expose MyModule and the method to JavaScript via
 * `NativeModules.MyModule.doSomething`
 */
#define RCT_EXTERN_MODULE(objc_name, objc_supername) RCT_EXTERN_REMAP_MODULE(, objc_name, objc_supername)
```

But I cannot see RCT_EXTERN_METHOD in this tutorial. Does that mean this method is not working any more in new arch?

**Comment by mateusz1913** [comment, marked as answer]
Hi @XuNeal, sorry for late response, I was on holidays. I moved it to Discussions tab, as it is more of a question rather than the issue with the tutorial.

The difference between old and new arch on iOS is that, new arch introduces additional communication layer written in C++. Because, Swift & C++ interoperability is very limited at the moment, we cannot directly call Swift classes, methods, functions, etc. with C++ arguments - which might be the case in new arch (e.g. when you want to pass a JS object argument into your native module method). You can take a look at the "official" samples in react native community github, where 
Swift code is used in the same way as in the tutorial (via importing Swift class into ObjC file and using explicitly):
- https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/turbomodule-swift
- https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/swift-event-emitter

If Swift & C++ interoperability will be improved at some point, the `RCT_EXTERN_` macros might be still valid, but for now, we have to explicitly use Swift in the ObjC++ files
